src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp

```diff
@@ -98,3 +104,100 @@ void G1FullGCCompactTask::serial_compaction() {
     compact_region(*it);
   }
 }
+
+void G1FullGCCompactTask::humongous_compaction() {
+  GCTraceTime(Debug, gc, phases) tm("Phase 4: Humonguous Compaction", collector()->scope()->timer());
+
+  for (HeapRegion* hr : *(collector()->humongous_compaction_regions())) {
+    assert(collector()->is_compaction_target(hr->hrm_index()), "Sanity");
+    compact_humongous_obj(hr);
+  }
+}
+
+void G1FullGCCompactTask::reset_humongous_metadata(HeapRegion* start_hr, uint num_regions, size_t word_size) {
```

Comment:
Maybe this could somehow be refactored with `G1CollectedHeap::humongous_obj_allocate_initialize_regions`; probably not easily.